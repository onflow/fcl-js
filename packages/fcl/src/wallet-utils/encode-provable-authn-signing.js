import {withPrefix} from "@onflow/util-address"
import {invariant} from "@onflow/util-invariant"
import {encode} from '@onflow/rlp'

const rightPaddedHexBuffer = (value, pad) =>
  Buffer.from(value.padEnd(pad * 2, 0), "hex")

export const encodeMessageForProvableAuthnSigning = (address, timestamp, appDomainTag = "") => {
    invariant(address, "Encode Message From Provable Authn Error: address must be defined")
    invariant(timestamp, "Encode Message From Provable Authn Error: timestamp must be defined")

    const USER_DOMAIN_TAG = rightPaddedHexBuffer(Buffer.from("FLOW-V0.0-user").toString("hex"), 32).toString("hex")
    const APP_DOMAIN_TAG = rightPaddedHexBuffer(Buffer.from(appDomainTag).toString("hex"), 32).toString("hex")

    return USER_DOMAIN_TAG + (appDomainTag ? 
        encode([APP_DOMAIN_TAG, withPrefix(address), timestamp]).toString("hex")
        :
        encode([withPrefix(address), timestamp]).toString("hex")
    )
}
