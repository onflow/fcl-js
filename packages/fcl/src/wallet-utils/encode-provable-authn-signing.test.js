import {encodeMessageForProvableAuthnSigning} from "./encode-provable-authn-signing.js"

const address = "0xABC123DEF456"
const timestamp = 1632179933495
const appTag = "FLOW-JS-SDK"

describe("encode provable authn", () => {
    test("encode provable authn with appTag", () => {
        const message = encodeMessageForProvableAuthnSigning(address, timestamp, appTag)

        expect(message).toMatchSnapshot()
    })

    test("encode provable authn without appTag", () => {
        const message = encodeMessageForProvableAuthnSigning(address, timestamp)

        expect(message).toMatchSnapshot()
    })
})
