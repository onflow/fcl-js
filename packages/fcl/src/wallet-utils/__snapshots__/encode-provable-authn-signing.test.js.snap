// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode provable authn encode provable authn with appTag 1`] = `"464c4f572d56302e302d75736572000000000000000000000000000000000000f850b8403436346334663537326434613533326435333434346230303030303030303030303030303030303030303030303030303030303030303030303030303030303086abc123def45686017c05815137"`;

exports[`encode provable authn encode provable authn without appTag 1`] = `"464c4f572d56302e302d75736572000000000000000000000000000000000000ce86abc123def45686017c05815137"`;
